src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp

```diff
@@ -0,0 +1,271 @@
+/****************************************************************************
+ * Copyright (c) 2017-2024 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_DETAILS_DISTRIBUTED_TREE_NEAREST_HELPERS_HPP
+#define ARBORX_DETAILS_DISTRIBUTED_TREE_NEAREST_HELPERS_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_Box.hpp>
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Point.hpp>
+#include <ArborX_Ray.hpp>
+#include <ArborX_Sphere.hpp>
+
+namespace ArborX
+{
+
+namespace Experimental
+{
+
+// Constrained callback is a callback that a user promises to:
+// - be not pure
+// - be allowed to be called on non-final results
+// - produce exactly one result for each match
+template <class Callback>
+struct ConstrainedDistributedNearestCallback
+{
+  Callback _callback;
+
+  template <class... Args>
+  KOKKOS_FUNCTION void operator()(Args &&...args) const
+  {
+    _callback((Args &&) args...);
+  }
+};
+
+template <class Callback>
+auto declare_callback_constrained(Callback const &callback)
+{
+  return ConstrainedDistributedNearestCallback<Callback>{callback};
+}
+
+} // namespace Experimental
+
+namespace Details
+{
+
+struct DefaultCallbackWithRank
+{
+  int _rank;
+
+  template <typename Predicate, typename Value, typename OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &, Value const &value,
+                                  OutputFunctor const &out) const
+  {
+    out({value, _rank});
+  }
+};
+
+template <class Callback>
+struct is_constrained_callback : std::false_type
+{};
+template <class Callback>
+struct is_constrained_callback<
+    Experimental::ConstrainedDistributedNearestCallback<Callback>>
+    : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
```

Comment:
It is non-obvious to me why we need a specialization for the default callback (haven't thought too hard about it) so it might be worth a comment.

Reply:
We need it because we use default callback internally when implementing a distributed nearest query without user provided callback.

Reply:
Added a comment.